src/main/scala/net/ripe/rpki/publicationserver/PublicationService.scala

```diff
@@ -112,13 +111,67 @@ class PublicationService
     }
 
   private def processRequest[T](parsedMessage: T, clientId: ClientId) = {
-    implicit val timeout: Timeout = Timeout(61.seconds)
-    parsedMessage match {
-      case queryMessage@QueryMessage(_) =>
-        stateActor ? RawMessage(queryMessage, clientId)
-      case ListMessage() =>
-        stateActor ? RawMessage(ListMessage(), clientId)
+    Future {
+      parsedMessage match {
+        case queryMessage@QueryMessage(_) =>
+          processQueryMessage(queryMessage, clientId)
+        case ListMessage() =>
+          // TODO ??? implement tags for list query
+          processListMessage(clientId, None)
+      }
+    }
+  }
+
+  private def processQueryMessage(queryMessage: QueryMessage, clientId: ClientId) = {
+    try {
+      implicit val m = metrics
+      objectStore.applyChanges(queryMessage, clientId)
+      triggerFlush()
+      ReplyMsg {
+        queryMessage.pdus.map {
+          case PublishQ(uri, tag, _, _) => PublishR(uri, tag)
+          case WithdrawQ(uri, tag, _) => WithdrawR(uri, tag)
+        }
+      }
+    } catch {
+      case e: RollbackException => ErrorMsg(e.error)
+      case e: Exception => ErrorMsg(BaseError("other_error", s"Unknown error: ${e.getMessage}"))
     }
   }
+
+  def processListMessage(clientId: ClientId, tag: Option[String]) =
+    ReplyMsg {
+      objectStore.list(clientId).map { case (url, hash) =>
+        ListR(URI.create(url), hash, tag)
+      }
+    }
+
+  var lastTimeFlushed : Option[Instant] = None
+
+  def triggerFlush() = {
+
+    def flush = synchronized {
+      dataFlusher.updateFS()
+      logger.info("Updated FS")
+      lastTimeFlushed = Some(Instant.now())
+    }
+
+    lastTimeFlushed match {
+      case None =>
+        flush
+      case Some(timeBefore) =>
+        val now = Instant.now()
+        val duration =
```

Comment:
```suggestion
          if (timeBefore.isBefore(now.minus(conf.snapshotSyncDelay.toSeconds, ChronoUnit.SECONDS))) {
            Duration(100, MILLISECONDS)
          } else {
            val between = (timeBefore.toEpochMilli - now.toEpochMilli)/1000
            val left = conf.snapshotSyncDelay.toSeconds - between
            Duration(left, SECONDS)
          }
```
Nit: Maybe use Scala Duration? Duration between java 8 and 11 has difference in the toSeconds, somehow this one was only compiled on java 11.